pulsar-client-cpp/lib/ClientConnection.cc

```diff
@@ -857,6 +857,20 @@ void ClientConnection::handleIncomingCommand() {
                     } else {
                         lock.unlock();
                     }
+
+                    Lock lock2(mutex_);
```

Comment:
I'd rather avoid the lock-unlock-lock-unlock. Instead of the first `else { lock.unlock() }` we could have the 2nd part and only unlock at the end, since if we do get an error response, it can only be of one single type.